backend/inspirehep/snow/api.py

```diff
@@ -0,0 +1,635 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# inspirehep is free software; you can redistribute it and/or modify it under
+# the terms of the MIT License; see LICENSE file for more details.
+
+import urllib.parse
+from functools import cached_property, wraps
+
+import backoff
+import orjson
+import requests
+import structlog
+from flask import current_app, render_template
+from inspire_utils.record import get_value
+from invenio_cache import current_cache
+
+from inspirehep.utils import DistributedLockError, distributed_lock
+
+from .errors import CreateTicketException, EditTicketException, SnowAuthenticationError
+from .utils import get_from_cache, get_response_result, strip_lines
+
+LOGGER = structlog.getLogger()
+
+
+class SnowTicketAPI:
+    """Implements basic methods to interact with SNOW (CERN implementation).
+
+    Attributes:
+        client_id (str): the id of the snow integration.
+        secret (str): the client secret.
+        auth_url (str): the url used to obtain the authentication token.
+        verify_cert (bool): is SSL certificate verification enabled.
+        url (str): the url of SNOW API.
+    """
+
+    def __init__(
+        self, url=None, client_id=None, secret=None, auth_url=None, verify_cert=True
+    ):
+        self.client_id = client_id or current_app.config.get("SNOW_CLIENT_ID")
+        self.secret = secret or current_app.config.get("SNOW_CLIENT_SECRET")
+        self.auth_url = auth_url or current_app.config.get("SNOW_AUTH_URL")
+        self.verify_cert = verify_cert
+        self.url = url or current_app.config.get("SNOW_URL")
+
+    def relogin_if_needed(f):
+        """Recalculate SNOW token if it's invalid."""
+
+        @wraps(f)
+        def decorated_function(self, *args, **kwargs):
+            try:
+                return f(self, *args, **kwargs)
+            except requests.exceptions.HTTPError:
+                if hasattr(self, "token"):
+                    del self.__dict__["token"]
+                return f(self, *args, **kwargs)
+
+        return decorated_function
+
+    @cached_property
+    def token(self):
+        token = current_cache.get("snow-token-value")
+        if not token:
+            token = self.get_token()
```

Comment:
why we have to run twice `get_token`? It's inside the `write_token_to_cache` as well